bson/src/main/org/bson/BSONCallbackAdapter.java

```diff
@@ -83,12 +86,21 @@ protected void doWriteEndArray() {
 
     @Override
     protected void doWriteBinaryData(final BsonBinary value) {
-        if (value.getType() == BsonBinarySubType.UUID_LEGACY.getValue()) {
-            bsonCallback.gotUUID(getName(),
-                    readLong(value.getData(), 0),
-                    readLong(value.getData(), 8));
+        UuidRepresentation defaultUuidRepresentation = getDefaultUuidRepresentation();
+        if (defaultUuidRepresentation == UuidRepresentation.STANDARD) {
+            if (value.getType() == BsonBinarySubType.UUID_LEGACY.getValue()) {
+                bsonCallback.gotBinary(getName(), value.getType(), value.getData());
+            } else {
+                UUID uuid = UuidHelper.decodeBinaryToUuid(value.getData(), value.getType(), defaultUuidRepresentation);
+                bsonCallback.gotUUID(getName(), uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
+            }
         } else {
-            bsonCallback.gotBinary(getName(), value.getType(), value.getData());
+            if (value.getType() == BsonBinarySubType.UUID_LEGACY.getValue()) {
+                UUID uuid = UuidHelper.decodeBinaryToUuid(value.getData(), value.getType(), defaultUuidRepresentation);
+                bsonCallback.gotUUID(getName(), uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
+            } else {
+                bsonCallback.gotBinary(getName(), value.getType(), value.getData());
+            }
```

Comment:
I cannot understand neither the old nor the new code. I took a pause and tried again in a day, but I still can't make any sense even of the old code. Jeff, could you please help me?

- The old code was decoding UUID values if they had the legacy format, and was omitting the decoding them if they were in the standard format. Why was it done this way, and how/when the values that had not been decoded were then decoded?
- The new code brings at least one more things that I fail to understand: the meaning of `BasicBSONDecoder.getDefaultUuidRepresentation`. The interpretation I can think of is "the format of the binary data that is being decoded". However, that format is specified in `value.getType`, and indeed, `BasicBSONDecoder.getDefaultUuidRepresentation` is not used as a specifier of the decoded data format. Instead, it is used to either decode the data as `UUID` or omit decoding it, and depending on the `getDefaultUuidRepresentation` value, the code omits decoding in different cases. Even looking at the `should decode UUID according to default uuid representation` test did not help me to understand what's happening.

Reply:
It's all part of the accidents of history...

In the beginning, there was only binary subtype 3 for UUIDs and a bunch of drivers spent their days happily decoding them into platform UUID objects (like java.util.UUID in Java) and encoding them back into binary subtype 3.

But lo, all was not well in the world of UUIDs.  It turned out that all the drivers twiddled the UUID bits in different ways when encoding them, and twiddled them back in the same way when decoding them.  So a binary subtype 3 sitting in a MongoDB document on some server could be decoding in three different ways (the java way, the C# way, and the python way).

And thus was born binary subtype 4.  A new subtype to rule them all.  For subtype 4 we picked the one true way to twiddle the bits (the python way, which is also the way it's specified in [the holy book](https://datatracker.ietf.org/doc/html/rfc4122#section-4.1.2).  

But lo (again, lo) that solved one problem but created more problems.  Now each UUID-supporting driver has to decide how to encode a UUID.  Should it use the unholy format that it had been using, or the new format?  And how to decode?  Should it turn all subtype 3 and 4 into platform UUIDs?  And if so, how to round-trip the UUIDs back to the database? A true mess.

So it was decided that the binary subtype 3's would continue to be decoded into platform UUIDs, and platform UUIDs would be encoding into binary subtype 3's.  And binary subtype 4 would be decoding into Binary, which no bit twiddling done at all.  And thus round-tripping was preserved.  But oh, still what a mess.

And that's how it remained until around 2019, when a sage engineer (Bernie Hackett) decided that we could no longer live in such confusion, and thus a [new holy text](https://github.com/mongodb/specifications/blob/master/source/uuid.rst) was born.  And the 4.0 Java driver obeyed the letter of this law, for the most part.  It turns out that there were a few corners of the legacy Java driver API that remained in the dark.  And here they are. 

So what is defaultUuidRepresentation and why does it do what it does?  Hopefully [this section](https://github.com/mongodb/specifications/blob/master/source/uuid.rst#implicit-decoding-and-encoding) will explain it.

Reply:
Nice writing style! Thank you for the detailed explanation and the link to the spec. Now it all makes sense to me, including the implicit decoding table from the spec. And the current code complies to that table.